### PR TITLE
[doxygen] Backport patch to enable modern compilers for 1.9.4

### LINKED ISSUE #16741: [package] doxygen/1.9.4 doesn't build with gcc 12 or clang 15
### Description

Looks like doxygen 1.9.4 won't build with gcc12/clang15 as per https://github.com/doxygen/doxygen/issues/9312. The error thrown is:

```
[ 49%] Building CXX object addon/doxywizard/CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o
cd /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/build/addon/doxywizard && /usr/bin/c++ -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_CAST_FROM_ASCII -DQT_NO_CAST_TO_ASCII -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_XML_LIB -DUNICODE -I/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/addon/doxywizard/. -I/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/libversion -I/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/build/generated_src -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtXml -O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 -fasynchronous-unwind-tables -DNDEBUG -fvisibility=hidden -fvisibility-inlines-hidden -fPIC -std=gnu++17 -MD -MT addon/doxywizard/CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o -MF CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o.d -o CMakeFiles/doxywizard.dir/moc_inputstrlist.cpp.o -c /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/build/addon/doxywizard/moc_inputstrlist.cpp
In file included from /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/doxygen.h:24,
                 from /home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/message.cpp:21:
/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/cache.h: In member function 'V* Cache<K, V>::insert(const K&, V&&)':
/home/iurt/rpmbuild/BUILD/doxygen-1.9.4/src/cache.h:53:14: error: 'exchange' is not a member of 'std'
   53 |         std::exchange(it->second->second,value);
      |              ^~~~~~~~
```

This isn't a problem with the recipe per se, and this is fixed in 1.9.5 so https://github.com/conan-io/conan-center-index/pull/16310 will provide a working solution.

The proposal for this ticket is to backport the patch since it's relatively trivial: https://github.com/doxygen/doxygen/commit/5198966c8d5fec89116d025c74934ac03ea511fa, which would give us a working doxygen 1.9.4 with more modern compilers.

### Package and Environment Details

NA

### Conan profile

```
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
[conf]
tools.build:compiler_executables={'c': '/usr/bin/gcc', 'cpp': '/usr/bin/g++'}
```

### Steps to reproduce

```
$ conan install --requires=doxygen/1.9.4 --build=missing
```

### Logs

<details><summary>Click to expand log</summary>

```
NA
```

</details>

### PR DESCRIPTION
**Pull request by samuel-emrys**
Closes #16741

Specify library name and version:  **doxygen/1.9.4**

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/doxygen/all/conandata.yml [hunk 1]
```diff
@@ -18,6 +18,10 @@ sources:
     url: "https://sourceforge.net/projects/doxygen/files/doxygen-1.8.17.src.tar.gz"
     sha256: "2cba988af2d495541cbbe5541b3bee0ee11144dcb23a81eada19f5501fd8b599"
 patches:
+  "1.9.4":
+    - patch_file: "patches/1.9.4-0001-enable-modern-compilers.patch"
+      patch_description: "Enable modern compilers"
+      patch_type: "bugfix"
```
Comment (SpaceIm) [inline recipes/doxygen/all/conandata.yml:24]:
```suggestion
      patch_type: "portability"
```

**Comment by SpaceIm** [comment]
Is it needed for other versions as well (if it's needed, they will break once c3i configuration updated to test gcc12 if this patch is not backported to these versions).

**Comment by samuel-emrys** [comment]
@SpaceIm not as far as I can tell. I built everything with clang 15 and it was just 1.9.4 that failed. I've only tested 1.9.4 and 1.9.2 with gcc12, but 1.9.2 was fine

File: recipes/doxygen/all/conandata.yml (hunk 1 above)
Comment (samuel-emrys) [reply recipes/doxygen/all/conandata.yml:24]: my read of https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/conandata_yml_format.md#L209 is that that `bugfix` would be more appropriate than `portability`

Comment (SpaceIm) [reply recipes/doxygen/all/conandata.yml:24]: bugfix is for bugs in code leading to runtime errors, while portability is everything related to build itself, that's my understanding.

Comment (samuel-emrys) [reply recipes/doxygen/all/conandata.yml:24]: Yeah I can squint at that interpretation.. aside from the build/runtime consideration, this description is the closest fit:
> Indicates a patch that backports an existing bug fix from the newer release or master branch (or equivalent, such as main/develop/trunk/etc).
@prince-chrismc @uilianries can either of you provide clarity on what patch type is most appropriate when a later bug fix affects a build?

Comment (prince-chrismc) [reply recipes/doxygen/all/conandata.yml:24]: Bugfix == library code implementation 
Portability == new platforms or build environments

File: recipes/doxygen/all/patches/1.9.4-0001-enable-modern-compilers.patch [hunk 2]
```diff
@@ -0,0 +1,25 @@
+From 5198966c8d5fec89116d025c74934ac03ea511fa Mon Sep 17 00:00:00 2001
+From: Dimitri van Heesch <doxygen@gmail.com>
```
Comment (prince-chrismc) [inline recipes/doxygen/all/patches/1.9.4-0001-enable-modern-compilers.patch:2]: You should add a patch source with a link to the PR or commit this is from please

**Review by SpaceIm — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

File: recipes/doxygen/all/conandata.yml (hunk 1 above)
Comment (prince-chrismc) [reply recipes/doxygen/all/conandata.yml]: Implementation... you are literally adding a header, that won't change the logic in any way.

It's portability here

**Comment by prince-chrismc** [comment]
Now it works

**Pull request merged**